### PR TITLE
Added some more instances for Eff

### PR DESCRIPTION
**Pull request by mmhat**
- `Monoid`
- `PrimMonad`
- `Semigroup`

### CODE REVIEW

**Comment by mmhat** [comment]
I defined `PrimMonad` in terms of the underlying `IO` monad. But we could instead use the `State` effect (or even a separate `PrimEffect`).

**Comment by arybczak** [comment]
`PrimMonad` is fine, but what do the instances for `Semigroup` and `Monoid` do?

**Comment by arybczak** [comment]
I wonder about gating the instance behind the `PrimE` effect and require `IOE` only for `runPrimE`.

In principle it sounds silly since PrimMonad allows you to go *lower* than IO, but on the other hand that would allow to use MutableByteArrayS etc. without full IO access. Well, at least on the surface, but it might still be useful separation as going back to IO requires you to get access to the IO constructor, which would be clearly visible in the code.

**Comment by mmhat** [comment]
I'd also prefer to have a separate effect for that but wanted to have your opinion on this first :+1:

File: src/Effectful/Internal/Monad.hs [hunk 1]
```diff
@@ -272,6 +279,21 @@ instance IOE :> es => MonadBaseControl IO (Eff es) where
...
   withRunInIO = unliftEff
 
+----------------------------------------
+-- Primitive
+
+-- | An effect to perform primitive state-transformer actions
+data PrimE :: Effect where
+  PrimE :: PrimE m r
+
+-- | Run an 'Eff' with primitive state-transformer actions
```
Comment (arybczak) [inline src/Effectful/Internal/Monad.hs:289]:
```suggestion
-- | Run an 'Eff' operation with primitive state-transformer actions.
```

File: src/Effectful/Internal/Monad.hs [hunk 2]
```diff
@@ -272,6 +279,21 @@ instance IOE :> es => MonadBaseControl IO (Eff es) where
 instance IOE :> es => MonadUnliftIO (Eff es) where
   withRunInIO = unliftEff
 
+----------------------------------------
+-- Primitive
+
+-- | An effect to perform primitive state-transformer actions
```
Comment (arybczak) [inline src/Effectful/Internal/Monad.hs:285]:
```suggestion
-- | An effect to perform primitive state-transformer actions.
```

File: src/Effectful/Internal/Monad.hs [hunk 3]
```diff
@@ -272,6 +279,21 @@ instance IOE :> es => MonadBaseControl IO (Eff es) where
...
+data PrimE :: Effect where
+  PrimE :: PrimE m r
+
+-- | Run an 'Eff' with primitive state-transformer actions
+runPrimE :: Eff (PrimE : es) a -> Eff es a
```
Comment (arybczak) [inline src/Effectful/Internal/Monad.hs:290]: This needs `IOE :> es` constraint.

Comment (mmhat) [reply src/Effectful/Internal/Monad.hs]: Done.

**Comment by arybczak** [comment]
Thanks :)

**Pull request merged**